src/lib/operators/join_sort_merge.cpp

```diff
@@ -497,26 +494,47 @@ class JoinSortMerge::JoinSortMergeImpl : public AbstractReadOnlyOperatorImpl {
         _sorted_right_table, [&](RowID right_row_id) { _emit_combination(output_cluster, NULL_ROW_ID, right_row_id); });
   }
 
-  /**
-  * Determines the length of the run starting at start_index in the values vector.
-  * A run is a series of the same value.
-  **/
-  size_t _run_length(size_t start_index, std::shared_ptr<MaterializedSegment<T>> values) {
-    if (start_index >= values->size()) {
+  // Determines the length of the run starting at start_index in the values vector. A run is a series of the same
+  // value. Even though the input vector is sorted, we first start by linearly scanning for the end of the run. The
+  // reason is that we know with a high probability that the item we are looking for is at the beginning of the input
+  // vector. If we do not find the run's end within the linearly scanned part, we use a binary search.
+  size_t _run_length(size_t start_index, const MaterializedSegment<T>& values) {
+    if (start_index >= values.size()) {
       return 0;
     }
 
-    auto start_position = values->begin() + start_index;
-    auto result = std::upper_bound(start_position, values->end(), *start_position,
-                                   [](const auto& a, const auto& b) { return a.value < b.value; });
+    const auto begin = values.begin() + start_index;
+    const auto run_value = begin->value;
+
+    constexpr auto LINEAR_SEARCH_ITEMS = size_t{128};
+    auto end = begin + LINEAR_SEARCH_ITEMS;
+    if (values.end() < end) {
```

Comment:
Is this a safe comparison? I don't know. Alternatively, we could check whether `start_index + LINEAR_SEARCH_ITEMS >= values.size()`